examples/siro_sandbox/controllers/fetch_baselines_controller.py

```diff
@@ -477,23 +512,34 @@ def act(self, obs, env):
                     )
                 type_of_skill = self.defined_skills.nav_to_robot.skill_name
 
-            if type_of_skill == "OracleNavPolicy":
-                finished_nav = obs["agent_0_has_finished_oracle_nav"]
+            # Check if we can safely terminate the follow policy
+            # We consider termination of FOLLOW state if
+            # (1) the human does not move, and
+            # (2) the robot is already close enough to the human
+            if (
+                not is_human_move
+                and human_robot_geo_dis < ROBOT_CLOSE_TO_HUMAN_DIS
+            ):
+                self._finished_follow = True
 
-            if type_of_skill == "NavSkillPolicy":
-                action_array = self.force_apply_skill(
-                    obs, target_pddl_skill, env, human_pos
-                )[0]
+            if not self._finished_follow:
+                if type_of_skill == "NavSkillPolicy":
+                    action_array = self.force_apply_skill(
+                        obs, target_pddl_skill, env, human_pos
+                    )[0]
 
-            elif type_of_skill == "OracleNavPolicy":
-                action_ind_nav = find_action_range(
-                    act_space, "agent_0_oracle_nav_action"
-                )
-                action_array[
-                    action_ind_nav[0] : action_ind_nav[0] + 3
-                ] = human_pos
+                elif type_of_skill == "OracleNavPolicy":
+                    action_ind_nav = find_action_range(
+                        act_space, "agent_0_oracle_nav_action"
+                    )
+                    action_array[
+                        action_ind_nav[0] : action_ind_nav[0] + 3
+                    ] = human_pos
+                else:
+                    raise ValueError(f"Skill {type_of_skill} not recognized.")
             else:
-                raise ValueError(f"Skill {type_of_skill} not recognized.")
+                self.current_state = FetchState.WAIT
```

Comment:
Given the current complexity of your state machine, can I suggest you get more organized about how to transition states? Just as a tiny example here of the kind of bugs we want to avoid, I wonder if you should have cleared `gt_path` on this transition right here. This kind of stuff is much easier to catch if you have one `change_state` helper function that lets you carefully think about every possible state transition, e.g. always calling `self._policy_info = self._get_policy_info()` and always clearing `gt_path`.

For reference, I had started to do that with my `cancel_fetch` helper function, although that was aimed more narrowly at ending the fetch behavior (not relevant for this state transition).

Reply:
Related: it looks like `act()` is over 500 lines long now... that suggests you want to do some cleanup to make it more readable and maintainable.

Reply:
Thanks, Eric! Working on this

Reply:
Initial implementation is done